### ISSUE TITLE
Flushing headers after last DATA write fails in DefaultHttp2ConnectionEncoder after window update

### ISSUE THREAD

**Issue by louiscryan** [netty/netty#3352]
To maintain HEADER frame order with respect to DATA frames the code listens to a ChannelFuture for the last DATA frame write. When DATA frames are queued and a window update occurs the DATA frames are written and flushed and the listener is called immediately in the event loop thread. This triggers and attempt to write and flush the header in the same event loop which fails as flush is not re-entrant.

I'm working on a patch the instead simply adds the headers to the same queue maintained from DATA frames inside the flow-controller. This should be more efficient in general as only one flush will be needed for combinations of DATA & HEADER frames. Will send pull request in a bit...

**Comment by normanmaurer** [comment]
@Scottmitch @nmittler FYI....

**Comment by Scottmitch** [comment]
> This triggers and attempt to write and flush the header in the same event loop which fails as flush is not re-entrant.

I see the potential for the issue you describe.  If the [AbstractChannel.doWrite](https://github.com/netty/netty/blob/master/transport/src/main/java/io/netty/channel/AbstractChannel.java#L799) call ends up invoking the [operationComplete](https://github.com/netty/netty/blob/master/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java#L272) for the pending header then the [flush](https://github.com/netty/netty/blob/master/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java#L297) for the pending header will not do what it is expected to do.  Not the most intuitive thing (flush not being re-entrant), but as you said we may be able to avoid this call pattern all together.

What you suggest sounds reasonable.  We will of-course have to account for the fact that HEADER frames don't count against flow control (and are not subject to flow control rules).

**Comment by Scottmitch** [comment]
@nmittler - WDTY.  It may not be ideal to have "non flow controlled frames" interact with the flow controller, but perhaps it makes sense to ensure proper sequencing?  We could add a method and remove `lastFlowControlledFrameSent`.  `Http2RemoteFlowController` may look something like the following:

```java
public interface Http2RemoteFlowController extends Http2FlowController {
/**
 * existing method
 */
ChannelFuture sendFlowControlledFrame(ChannelHandlerContext ctx, Http2Stream stream,
            ByteBuf data, int padding, boolean endStream, ChannelPromise promise);

/**
 * new method which doesn't count against flow control, but ensures proper sequencing with other non-flow controlled and flow-controlled frames this class has seen.
 */
ChannelFuture sendNonFlowControlledFrame(ChannelHandlerContext ctx, Http2Stream stream,
            ByteBuf data, int padding, boolean endStream, ChannelPromise promise);
}

**Comment by nmittler** [comment]
@Scottmitch I think this is probably the best way to handle it.  What we had was admittedly hacky and in need of a face lift :)

**Comment by louiscryan** [comment]
Pull request is basically along those lines though I limited it to Headers. I don't think it's useful to queue PRIORITY or WINDOW_UPDATE as they don't affect the data visible to the application layer. If you care about the semantics of when RST_STREAM breaks your data stream then you might care but that seems of very little value.  

There aren't any other frames in the spec that you might care about but I can see this becoming necessary for extension frames when they are supported.

**Comment by Scottmitch** [comment]
> Pull request is basically along those lines though I limited it to Headers

Yes that makes sense.  The interface I described for `sendNonFlowControlledFrame` didn't make much sense because data should always be flow controlled....but as long as you get the idea (just substitute the data parts for `Http2Header` parts) :).

> There aren't any other frames in the spec that you might care about but I can see this becoming necessary for extension frames when they are supported.

Agreed.  We do have https://github.com/netty/netty/issues/3222 opened for extension frames, but we can deal with that when the time comes.

**Comment by Scottmitch** [comment]
Link to PR https://github.com/netty/netty/pull/3353

**Comment by nmittler** [comment]
Closing since this was fixed by #3353
